pkg/proxy/userspace/proxysocket.go

```diff
@@ -104,9 +115,19 @@ func tryConnect(service proxy.ServicePortName, srcAddr net.Addr, protocol string
 			glog.Errorf("Dial failed: %v", err)
 			continue
 		}
-		return outConn, nil
+		break
 	}
-	return nil, fmt.Errorf("failed to connect to an endpoint.")
+	if outConn == nil && err == nil {
+		err = fmt.Errorf("failed to connect to an endpoint.")
+	}
+	if err != nil {
+		if protocol == "tcp" {
```

Comment:
This is a little smelly to me.  The main reason you can't just call inConn.Close() is that you're passing the UDP listener as the inConn.  I think a better pattern is something like this, tell me what you think:

in here:

```
if err != nil {
    glog.Errorf("Failed to connect to balancer: %v", err)
    inConn.Close()
    return
}
```

in UDP code:

```
// fakeUDPConn pretends to be a "connection" to a UDP client so that other code can safely call Close().
// Of course, UDP doesn't really have a connection, so just fake it.
type fakeUDPConn struct {
    net.Conn
}

func (u *fakeUDPConnection) Close() {
    // Do nothing
}
```

In UDP connectBackend:

```
go connectAndProxy(service, fakeUDPConnection{udp}, cliAddr, "udp", proxier, func(in, svrConn net.Conn) {
   // everything else
}
```

Another alternative would be to pass a closeInput function to connectAndProxy() and make TCP's be `inConn.Close()` and UDP's be an empty function.